File: Next Greater Element I/Next_Greater_Element_I.cpp
// Runtime: 12 ms, faster than 75.78% of C++ online submissions for Next Greater Element I.
// Memory Usage: 9.3 MB, less than 65.43% of C++ online submissions for Next Greater Element I.

class Solution 
{
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) 
    {
        
        unordered_map<int, int> hashMap;
        stack<int> myStack;

        for (int index = nums2.size() - 1; index >= 0 ; index--)
        {
            while (!myStack.empty() && nums2[index] >= myStack.top())
                myStack.pop();
            
            hashMap.insert(make_pair(nums2[index], myStack.empty() ? -1 : myStack.top()));
            myStack.push(nums2[index]);
        }

        vector<int> res(nums1.size());
        for (int index = 0; index < nums1.size(); index++)
            res[index] = hashMap.at(nums1[index]);

        return res;
    }
};

File: Next Greater Element I/Next_Greater_Element_I.py
# Runtime: 40 ms, faster than 94.25% of Python3 online submissions for Next Greater Element I.
# Memory Usage: 13.2 MB, less than 85.02% of Python3 online submissions for Next Greater Element I.

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        hashMap = dict()
        stack = []

        for i in range(len(nums2) - 1, -1, -1):
            while myStack and nums2[i] >= myStack[-1]:
                myStack.pop()
            
            hashMap[nums2[index]] = -1 if not myStack else myStack[-1]
            myStack.push(nums2[i])
        
        res = []
        for item in nums1:
            res.append(hashMap[item])
        return res